javascript/ql/src/experimental/Security/CWE-020/PostMessageNoOriginCheck.ql

```diff
@@ -0,0 +1,71 @@
+/**
+ * @name Missing `MessageEvent.origin` verification in `postMessage` handlers
+ * @description Missing the `MessageEvent.origin` verification in `postMessage` handlers, allows any windows to send arbitrary data to the `MessageEvent` listener.
+ * This could lead to unexpected behaviour, especially when `MessageEvent.data` is used in an unsafe way.
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @id js/missing-postmessageorigin-verification
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-20
+ */
+
+import javascript
+import semmle.javascript.security.dataflow.DOM
+
+/**
+ * A method call for the insecure functions used to verify the `MessageEvent.origin`.
+ */
+class InsufficientOriginChecks extends DataFlow::MethodCallNode {
```

Comment:
We have a few API methods for calls like this already. Do they suit your needs?

See this for instance: https://github.com/github/codeql/blob/d80a033bed070c98ce91b74986098297e5e11ab1/javascript/ql/src/Security/CWE-020/IncompleteUrlSubstringSanitization.ql#L16-L32

Reply:
Thanks for pointing this out.  I was already aware of these APIs, indeed I was inspired by a similar query:

[codeql/javascript/ql/src/Security/CWE-020/IncorrectSuffixCheck.ql#L18-L23](https://github.com/github/codeql/blob/d80a033bed070c98ce91b74986098297e5e11ab1/javascript/ql/src/Security/CWE-020/IncorrectSuffixCheck.ql#L18-L23)

Since I am pretty new to this framework, what are the benefits of using the suggested API methods?

I just want to detect if there is weak function call (like `indexOf`) to validate the origin.

It seems to me that these API methods are preferred if you what to reason about the object passed to these methods.

In any case, if you think that the actual implementation can be improved, let me know. I will try to combine the already existing API methods and what I have done.

Again, thanks for your feedback.

Reply:
This is just a minor suggestion, feel free to ignore it. 

> Since I am pretty new to this framework, what are the benefits of using the suggested API methods?

The benefit is that `StringOps::...` is like to cover more cases, and that it is likely to be extended in the future. See this case for instance:

https://github.com/github/codeql/blob/03c6d7a7e5870aa653a04d869f2f83c7b697a6e0/javascript/ql/src/semmle/javascript/InclusionTests.qll#L110-L122

Of course, if you spot a false positive from your query due to the use of `StringOps::...`, then the right choice is the simpler list of method names that you already have.

Reply:
Thanks for explaining this :)


Glad to know the difference. Since `StringOps` covers more cases of functions that can be used to validate the origin (improperly), I changed the code in favour of `StringOps`.
Thanks for the feedback.